src/boot/bootctl.c

```diff
@@ -2388,6 +2459,25 @@ static int run(int argc, char *argv[]) {
         if (r <= 0)
                 return r;
 
+        /* Open up and mount the image */
+        if (arg_image) {
+                assert(!arg_root);
+
+                r = mount_image_privately_interactively(
+                                arg_image,
+                                DISSECT_IMAGE_GENERIC_ROOT |
+                                DISSECT_IMAGE_RELAX_VAR_CHECK,
+                                &unlink_dir,
+                                &loop_device,
+                                &decrypted_image);
```

Comment:
they ara marked autoclear, so should be cleaned up automatically.

Reply:
At least `_cleanup_(unlink_tempfilep)` doesn't work when programs where it's used are interrupted with Ctrl-C so I'm not sure why `_cleanup_(loop_device_unrefp)` should be any different.

Reply:
I ran `test-loop-block` (where `_cleanup_(loop_device_unrefp)` is used), interrupted it with CTRL-C and got a bunch of `/dev/loop*` laying around:
```sh
$ ls -l /dev/loop*
brw-rw----. 1 root disk  7,   0 Jul  6 16:32 /dev/loop0
brw-rw----. 1 root disk  7,   1 Jul  6 16:32 /dev/loop1
brw-rw----. 1 root disk  7,   2 Jul  6 16:32 /dev/loop2
brw-rw----. 1 root disk  7,   3 Jul  6 16:32 /dev/loop3
brw-------. 1 root root  7,   4 Jul  6 16:32 /dev/loop4
crw-rw----. 1 root disk 10, 237 Jul  6 12:25 /dev/loop-control
```
so it seems to me that loop devices are cleaned up when they are set up and managed by PID1. The command line tools can't expect systemd to clean them up.

Reply:
> I'm not sure why _cleanup_(loop_device_unrefp) should be any different.

we set the autoclear flag of the loopback device. The kernel will then free the device once it is not referenced anymore, i.e. not mounted or otherwise opened.

Reply:
> I ran test-loop-block (where _cleanup_(loop_device_unrefp) is used), interrupted it with CTRL-C and got a bunch of /dev/loop* lying around:

That's misleading. autoclean means the stuff is detached, but the devices stay around. but that's fine, as the next user will use them then. For the next process needing a loopback device there's no difference between "loopback device is gone", and "loopback device exists but not attached" the api calls are the same to allocate/attach something.

Allocation always happens from the bottom, so the worst that happens is that there might be one detached loopback device around for a bit until it is recycled.

Reply:
i mean, the API of loopback devices is terrible in a million ways, and the fact these devices stay around in detached state is pretty awful, but it's not a total loss. I wish they'd clean this up more comprehensively.